findatastructures/src/finset/mod.rs

```diff
@@ -0,0 +1,29 @@
+pub trait FinSet: PartialEq + Clone {
+    /// The type of values in the set
+    type V;
+
+    /// Create a new empty set
+    fn new() -> Self;
+    /// Check if the set contains a value
+    fn contains(&self, v: &Self::V) -> bool;
+    /// Insert a value into the set
+    fn insert(&self, v: Self::V) -> Self;
```

Comment:
Similar comments to the map. What's the design rational for these functions, esp. in deviating from the common behaviour in Rust?

Reply:
if its about not mutating the struct, again its due to the ensures macro